packages/botonic-nlp/src/codify/codifier.ts

```diff
@@ -0,0 +1,39 @@
+import { unique } from '../utils/array-utils'
+
+export class Codifier {
+  vocabulary: string[]
+
+  constructor(vocabulary: string[]) {
+    this.vocabulary = unique(vocabulary)
+  }
+
+  encode(sequence: string[]): number[] {
+    return sequence.map(token => this.encodeToken(token))
+  }
+
+  private encodeToken(token: string): number {
+    if (!this.isValidToken(token)) {
+      throw new Error(`Invalid Token '${token}'.`)
+    }
+    return this.vocabulary.indexOf(token)
+  }
+
+  private isValidToken(token: string): boolean {
+    return this.vocabulary.includes(token)
+  }
+
+  decode(sequence: number[]): string[] {
+    return sequence.map(id => this.decodeTokenId(id))
+  }
+
+  private decodeTokenId(id: number): string {
+    if (!this.isValidTokenId(id)) {
+      throw new Error(`Invalid Token id '${id}'.`)
+    }
+    return this.vocabulary[id]
+  }
+
+  private isValidTokenId(id: number) {
+    return this.vocabulary.length > id && id >= 0
```

Comment:
readibility: in a comparison, you should try to put the checked variable first  
id >=0 && id <=this.voca....
as this is how people would speak: you wouldn't say that you're checking that the vocabulary length is larger than the specified number, but that "you're checking that a number is smaller than the voc length"
another option would be 
 0<=id && id <=this.voca....
since it's simmetrical and it's easier to read left to right.